docs/utxo-hd/v0.1/for-developers.md

```diff
@@ -0,0 +1,276 @@
+boundary. For this reason, the general scheme for **ticking a block is: a ledger
+state without a UTxO set (`EmptyMK`) goes in and a ledger state with some deltas
+(`DiffMK`) comes out**. This is reflected in `applyChainTickLedgerResult`.
+
+```haskell
+applyChainTickLedgerResult ::
+     LedgerCfg l
+  -> SlotNo
+  -> l EmptyMK
+  -> LedgerResult l (Ticked1 l DiffMK)
+```
+
+The exceptions we refer to are:
+
+- For simplicity, we have deliberately chosen that the Byron era would not have
+  any ledger tables, and there would only be ledger tables starting in Shelley.
+  This means that when crossing the boundary to Shelley, the whole UTxO set will
+  be returned as deltas (see `translateLedgerStateByronToShelleyWrapper`
+  [here](../../../ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/CanHardFork.hs)).
+
+- In the Shelley to Allegra transition, a bunch of UTxOs (owned by the AVVM
+  addresses) were consumed into the treasury. This transition will return said
+  UTxOs as deletions (see `translateLedgerStateShelleyToAllegraWrapper`
+  [here](../../../ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/CanHardFork.hs))..
+
+### The mempool
+
+The mempool behaves pretty much as a virtual block. The design is not
+particularly complex as we just ask the backing store for the inputs for a
+transaction when applying it. The only caveat compared to the pre-UTxO-HD
+implementation is that previously it cached the latest ledger state and
+therefore we could run a separate thread that would sync the mempool with the
+LedgerDB and revalidate all the transactions when necessary, but now we might
+not be able to apply a transaction if the ledger state on top of which we had
+applied the others is now gone from the LedgerDB. Therefore, adding a
+transaction might in some cases trigger a sync with the LedgerDB (and
+revalidation).
+
+### Ledger state queries
+
+Most of the queries don't require the UTxO set but there are three in particular
+that do: `GetUTxOByTxIn`, `GetUTxOWhole` and `GetUTxOByAddress`. We assume that
+`GetUTxOWhole` is a debug query so we don't pay much attention to it as we
+expect it will be used with small-ish UTxO sets. For `GetUTxOByTxIn`, the story
+is simple, query the Backing store for the `TxIn` and forward it through the
+sequence of diffs.
+
+However, it is `GetUTxOByAddress` that poses a real problem, as we need to query
+the whole UTxO set, apply all the differences to it and then traverse it
+entirely to find out the UTxOs belonging to an address. This query is now quite
+slow and in fact its usage is discouraged. It should not be a responsibility of
+the node to maintain access to this if it is not needed by the logic that runs
+the blockchain, so the plan is to move this into a separate process/client that
+runs an index of UTxOs by address that can provide fast access to it.
```

Comment:
We might want to link again to https://github.com/input-output-hk/cardano-node/issues/4678